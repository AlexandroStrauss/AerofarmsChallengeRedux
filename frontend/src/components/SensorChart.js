import React from 'react';
import { useTranslation } from 'react-i18next';

import {
    LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer
} from 'recharts';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';


// query GraphQL, with the array of sensor IDs whose data we want passed up
const SENSOR_QUERY = gql`
    query ($data_type: String!, $ids: [ID]!) {
        dataPointsByType(data_type: $data_type, ids: $ids) {
		    id
            date
        	value
        }
    }
`;

function SensorChart({ ids, data_type }) {
    const { t } = useTranslation();

    const { loading, error, data } = useQuery(SENSOR_QUERY, {
        variables: { ids: ids, data_type: data_type }
    });


    if (loading) return <p>{t('loading')}</p>;
    if (error) return <p>{t('error')}</p>;
    return (

        // Recharts does the heavy lifting of creating the chart.
        <ResponsiveContainer width="100%" height={600} >
        <LineChart
            // width={700}
            // height={400}
            data={data.dataPointsByType}
            margin={{
                top: 5, right: 30, left: 80, bottom: 10,
            }}
        >
            <CartesianGrid strokeDasharray="3 3" />
            <XAxis dataKey="date" label={{ value: "time", dy: 15 }} />
            <YAxis label={{ value: data_type === "temperature" ? t("temperature") : t("humidity"), dx: -50 }} domain={[45, 85]}/> {/* I adjusted the range of the Y-axis to better suit the data. */}
            <Tooltip /> {/* The Tooltip functionality lets you mouse over the chart and see each data point. */}
            <Line type="monotone" dataKey="value" dot={false} stroke={data_type === "temperature" ? "#ff0000" : "#8884d8"} activeDot={{ r: 1 }} />{/* dot={false} keeps every data point from being displayed as a dot on the chart. Trust me, it's better this way.  */}
        </LineChart>
        </ResponsiveContainer>
    );
}

export default SensorChart;