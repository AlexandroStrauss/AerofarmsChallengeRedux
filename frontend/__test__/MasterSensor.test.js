import React, { Suspense, lazy } from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/react-testing';
import 'regenerator-runtime/runtime';

import toJson from 'enzyme-to-json';
// import MasterSensor from '../src/components/MasterSensorChart';
import { I18nextProvider } from 'react-i18next';
import i18n from '../src/i18n';
import SensorChart from '../src/components/SensorChart';
const MasterSensor = lazy(() => import('../src/components/MasterSensorChart'))

jest.mock('react-i18next', () => ({
    // this mock makes sure any components using the translate HoC receive the t function as a prop
    translate: () => Component => props => <Component t={() => ''} {...props} />,
  }));
  
function createNodeMock() {
    const doc = document.implementation.createHTMLDocument();
    return { parentElement: doc.body };
  }
  
const wait = require('waait');
// import 'core-js/stable';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../src/components/SensorChart', () => 'SensorChart');

// afterEach(cleanup);

describe('Rendering SensorChart', () => {
    it('should render SensorChart component', () => {

        const tree = Enzyme.mount(
            <Suspense fallback="ugh we still doing this huh">
                <MasterSensor />
                </Suspense>
        );

        // await renderer.act(async () => {
        //     await wait(0); // wait for response
        // }); 

        expect(tree.contains("data type")).toBe(true);

        // expect(toJson(tree)).toMatchSnapshot();
    })


    it('should render appropriate things', async () => {
        const options = {createNodeMock}

        // let component;
        const component = renderer.create(
            <Suspense fallback="looolol">

                <MockedProvider addTypename={false}>
                {/* // <I18nextProvider i18n={i18n}> */}
                    <MasterSensor />
                {/* // </I18nextProvider> */}
                </MockedProvider >
                </Suspense>, options
            );

        await renderer.act(async () => {
            await wait(0); // wait for response
        }); 

        // console.log(component); 

        const tree = component.root.toJSON();

        // console.log(tree);

        //this doesn't feel right; feels waaaaay too specific to find a specific string in the page
        expect(tree.children[1].children[0].children.join('')).toContain('Displaying temperature data for...')

    })

    it('should change to humidity display when you hit the button', async () => {
        const options = {createNodeMock}

        // renderer.act(() => {
        const component = renderer.create(
                <MockedProvider addTypename={false}>
                    <Suspense fallback="loading">
                        <I18nextProvider i18n={i18n}>


                    <MasterSensor />
                    </I18nextProvider>
                    </Suspense>
                </MockedProvider >, options
            );
        // })

        await renderer.act(async () => {
            await wait(0); // wait for response

        });


        // THIS WORKS! This finds the right button, clicks it, waits for state to change, then checks for updated state
        const root = component.root;

        // console.log(root);

        const button = root.findAllByType('button')[1];

        button.props.onClick();

        await renderer.act(async () => {
            await wait(0); // wait for response

        });


        const tree = component.toJSON();
        expect(tree.children[1].children[0].children.join('')).toContain('Displaying humidity data for...')

    })
})

it('should work', async () => {
    const options = {createNodeMock}
    console.log(MasterSensor.WrappedComponent);
    const tree = renderer.create(
        // <Suspense fallback={<div>Loading</div>}>
        <MasterSensor />
        // </Suspense>, options
    );

    // await import('../src/components/SensorChart');

    await renderer.act(async () => {
        await wait(0); // wait for response
    });

    // await wait(0);
    // await tree.getInstance().loadingPromise;

    console.log(tree.toJSON());
    expect(true).toBe(true); 
})