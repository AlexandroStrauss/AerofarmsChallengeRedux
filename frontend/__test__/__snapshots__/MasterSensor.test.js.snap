// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering SensorChart should render SensorChart component 1`] = `undefined`;
